File: TaskScheduler.swift
//
//  TaskScheduler.swift
//  TheS30-SwiftTests
//
//  Created by Malav Soni on 05/01/25.
//

import Testing

// MARK: - Task Scheduler Logic
class Task {
    var char: String
    var freq: Int

    init(char: String, freq: Int) {
        self.char = char
        self.freq = freq
    }
}

func leastIntervalWithHeap(_ tasks: [String], _ n: Int) -> Int {
    var freqMap: [String: Int] = [:]
    let heap = Heap<Task>(comparator: { $0.freq > $1.freq })

    for task in tasks {
        freqMap[task, default: 0] += 1
    }

    for (char, freq) in freqMap {
        heap.insert(Task(char: char, freq: freq))
    }

    var taskCount = 0

    while !heap.isEmpty {
        var cycle = n + 1
        var poppedTasks: [Task] = []

        while cycle > 0, let task = heap.poll() {
            task.freq -= 1
            if task.freq > 0 {
                poppedTasks.append(task)
            }
            taskCount += 1
            cycle -= 1
        }

        for task in poppedTasks {
            heap.insert(task)
        }

        if !heap.isEmpty {
            taskCount += cycle
        }
    }

    return taskCount
}

func leastInterval(_ tasks: [String], _ n: Int) -> Int {
    var freqMap: [String: Int] = [:]
    var maxFreq = 0
    var maxCount = 0

    for task in tasks {
        let freq = (freqMap[task] ?? 0) + 1
        freqMap[task] = freq
        maxFreq = max(maxFreq, freq)
    }

    for freq in freqMap.values where freq == maxFreq {
        maxCount += 1
    }

    let partitions = maxFreq - 1
    let availableSlots = partitions * (n - (maxCount - 1))
    let pendingTasks = tasks.count - maxFreq * maxCount
    let idleSlots = max(0, availableSlots - pendingTasks)

    return tasks.count + idleSlots
}

struct TaskSchedulerTests {
    
    struct TestCase {
        let intervals: [String]
        let n: Int
        let expected: Int
    }
    
    static let testCases: [TestCase] = [
        TestCase(intervals: ["A", "A", "A", "B", "B", "B"], n: 2, expected: 8),
        TestCase(intervals: ["A", "A", "A", "B", "B", "B"], n: 3, expected: 10)
    ]

    @Test("Greedy Logic", arguments: testCases)
    func greedyLogicTest(arguments: TestCase) {
        let leastIntervalResult = leastInterval(arguments.intervals, arguments.n)
        #expect(leastIntervalResult == arguments.expected)
    }
    
    @Test("Heap Logic", arguments: testCases)
    func heapLogic(arguments: TestCase) {
        let leastIntervalResult = leastIntervalWithHeap(arguments.intervals, arguments.n)
        #expect(leastIntervalResult == arguments.expected)
    }
}

File: task_scheduler.ts
import { Heap } from "../../helpers/heap_helper";

class Task {
  char: string;
  freq: number;
  constructor(char: string, freq: number) {
    this.char = char;
    this.freq = freq;
  }
}
// O(m log n)
function leastInterval_with_heap(tasks: string[], n: number): number {
  let freqMap: Map<string, number> = new Map();
  let heap = new Heap<Task>((a, b) => b.freq - a.freq);

  for (const task of tasks) {
    let freq = (freqMap.get(task) || 0) + 1;
    freqMap.set(task, freq);
  }

  for (const item of freqMap) {
    heap.insert(new Task(item[0], item[1]));
  }

  let path: string = "";
  let taskCount: number = 0;

  while (!heap.isEmpty()) {
    let cycle = n + 1;
    let poppedTask: Task[] = [];
    while (cycle > 0 && !heap.isEmpty()) {
      let popped = heap.poll()!;
      popped.freq -= 1;
      if (popped.freq > 0) {
        poppedTask.push(popped);
      }
      path += popped.char;
      cycle--;
      taskCount++;
    }

    for (const task of poppedTask) {
      heap.insert(task);
    }

    while (cycle > 0 && !heap.isEmpty()) {
      path += "_";
      cycle--;
      taskCount++;
    }
  }
  return taskCount;
}

function leastInterval(tasks: string[], n: number): number {
  let maxFreq = 0;
  let maxCount = 0;
  let freqMap: Map<string, number> = new Map();

  for (const task of tasks) {
    let freq = (freqMap.get(task) || 0) + 1;
    freqMap.set(task, freq);
    maxFreq = Math.max(maxFreq, freq);
  }

  for (const item of freqMap) {
    if (item[1] == maxFreq) maxCount++;
  }

  let partition = maxFreq - 1;
  let availableSlot = partition * (n - (maxCount - 1));
  let pendingTask = tasks.length - maxFreq * maxCount;
  let idle = Math.max(0, availableSlot - pendingTask);
  return tasks.length + idle;
}

describe("621. Task Scheduler", () => {
  it("Happy Path - 01", () => {
    expect(leastInterval(["A", "A", "A", "B", "B", "B"], 2)).toEqual(8);
  });

  it("Happy Path - With Heap Logic", () => {
    expect(leastInterval_with_heap(["A", "A", "A", "B", "B", "B"], 2)).toEqual(
      8
    );
  });

  it("Happy Path - With Heap Logic With N = 3", () => {
    expect(leastInterval_with_heap(["A", "A", "A", "B", "B", "B"], 3)).toEqual(
      10
    );
  });
});
